File: app/src/main/java/net/squanchy/eventdetails/domain/view/ExperienceLevel.java
package net.squanchy.eventdetails.domain.view;

import android.support.annotation.StringRes;

import java.util.Locale;

import net.squanchy.R;
import net.squanchy.support.lang.Optional;

import timber.log.Timber;

public enum ExperienceLevel {
    BEGINNER("beginner", R.string.experience_level_beginner),
    INTERMEDIATE("intermediate", R.string.experience_level_intermediate),
    ADVANCED("advanced", R.string.experience_level_advanced);

    private final String rawLevel;

    private final int labelStringResId;
    ExperienceLevel(String rawLevel, @StringRes int labelStringResId) {

        this.rawLevel = rawLevel;
        this.labelStringResId = labelStringResId;
    }

    public String rawLevel() {
        return rawLevel;
    }

    @StringRes
    public int labelStringResId() {
        return labelStringResId;
    }

    public static Optional<ExperienceLevel> fromNullableRawLevel(String rawLevel) {
        try {
            return Optional.fromNullable(rawLevel).map(ExperienceLevel::fromRawLevel);
        } catch (IllegalArgumentException e) {
            Timber.d(e);
            return Optional.absent();
        }
    }

    public static ExperienceLevel fromRawLevel(String rawLevel) {
        switch (rawLevel.toLowerCase(Locale.US)) {
            case "beginner":
                return ExperienceLevel.BEGINNER;
            case "intermediate":
                return ExperienceLevel.INTERMEDIATE;
            case "advanced":
                return ExperienceLevel.ADVANCED;
            default:
                throw new IllegalArgumentException("Invalid raw level description: " + rawLevel);
        }
    }
}

File: app/src/main/java/net/squanchy/eventdetails/domain/view/ExperienceLevel.kt
package net.squanchy.eventdetails.domain.view

import androidx.annotation.ColorRes
import androidx.annotation.StringRes
import arrow.core.Option
import net.squanchy.R
import timber.log.Timber
import java.util.Locale

enum class ExperienceLevel(
    private val rawLevel: String,
    @StringRes val labelStringResId: Int,
    @ColorRes val colorResId: Int
) {
    BEGINNER("beginner", R.string.experience_level_beginner, R.color.experience_level_beginner),
    INTERMEDIATE("intermediate", R.string.experience_level_intermediate, R.color.experience_level_intermediate),
    ADVANCED("advanced", R.string.experience_level_advanced, R.color.experience_level_advanced);

    companion object {

        fun tryParsingFrom(rawLevel: String?): Option<ExperienceLevel> =
            rawLevel?.let {
                try {
                    Option(fromRawLevel(it))
                } catch (e: IllegalArgumentException) {
                    Timber.d(e)
                    null
                }
            } ?: Option.empty()

        private fun fromRawLevel(rawLevel: String) =
            when (rawLevel.toLowerCase(Locale.US)) {
                BEGINNER.rawLevel -> BEGINNER
                INTERMEDIATE.rawLevel -> INTERMEDIATE
                ADVANCED.rawLevel -> ADVANCED
                else -> throw IllegalArgumentException("Invalid raw level description: $rawLevel")
            }
    }
}
